lib/shopify_api/bulk/query.ex

```diff
@@ -97,6 +98,51 @@ defmodule ShopifyAPI.Bulk.Query do
     end
   end
 
+  @spec stream_fetch(String.t() | {}) :: Enumerable.t()
+  def stream_fetch({:ok, url}), do: stream_fetch(url)
+  def stream_fetch({:error, _} = error), do: error
+
+  def stream_fetch(url) do
+    url
+    |> httpoison_streamed_get()
+    # Our http chunks are not guaranteed to be on the line break, this splits up our
+    # returned chunks by line break.
+    |> Stream.chunk_while(
+      [],
+      fn element, acc ->
+        element = "#{List.first(acc)}" <> element
+
+        cond do
+          # A nicely formatted jsonl line(s)
+          # {\"test\":\"bar\"}\n{\"test\":\"baz\"}\n"
+          String.ends_with?(element, "\n") ->
+            {:cont, String.split(element, "\n", trim: true), []}
+
+          # Contains a jsonl line and then some
+          # {\"test\":\"bar\"}\n{\"test\"
+          String.contains?(element, "\n") ->
+            {rem, e} =
+              element
+              |> String.split("\n", trim: true)
+              |> List.pop_at(-1)
+
+            {:cont, e, [rem]}
+
+          # No complete jsonl document yet
+          true ->
+            {:cont, [element]}
+        end
+      end,
+      fn
+        [] -> {:cont, []}
+        acc -> {:cont, acc, []}
+      end
+    )
+    # We will have a "list" of "lists of strings" here, we'll flatten it down to a single list
+    # [["some json", "more jsonl"], ["jsonl"]] => ["some json", "more jsonl", "jsonl"]
+    |> Stream.flat_map(& &1)
```

Comment:
The need to use `flat_map` here suggests to me that there's a slightly better function to use than `Stream.chunk_while/3` -- I believe we could implement basically the same logic using `Stream.transform/3`, which looks to be better designed to buffer/split an input stream into a different set of output elements.

Reply:
I'm also curious if this is where we should `Stream.map` the resulting lines through a JSON decoder.

Reply:
I was digging at transform this morning and may have that refactor up later 👍 A good suggestion!

I've been pondering where to put the json decoding as well. I was also toying with the idea of pulling all the fetching (httpoison) related code in to its own module in bulk.

Reply:
The thing to be mindful of with `Stream.transform` is that it _does not_ yield the accumulator when you halt. I recall being bitten by that before.